adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -0,0 +1,636 @@
+    protected void customizeDownstreamMessage(final Message downstreamMessage, final RoutingContext ctx) {
+        MessageHelper.addProperty(downstreamMessage, LoraConstants.APP_PROPERTY_ORIG_LORA_PROVIDER,
+                ctx.get(LoraConstants.APP_PROPERTY_ORIG_LORA_PROVIDER));
+    }
+
+    private void setupAuthorization(final Router router) {
+        final ChainAuthHandler authHandler = ChainAuthHandler.create();
+        authHandler.append(new X509AuthHandler(
+                new TenantServiceBasedX509Authentication(getTenantServiceClient(), tracer),
+                Optional.ofNullable(clientCertAuthProvider).orElse(
+                        new X509AuthProvider(getCredentialsServiceClient(), getConfig(), tracer))));
+        authHandler.append(new HonoBasicAuthHandler(
+                Optional.ofNullable(usernamePasswordAuthProvider).orElse(
+                        new UsernamePasswordAuthProvider(getCredentialsServiceClient(), getConfig(), tracer)),
+                getConfig().getRealm(), tracer));
+
+        router.route().handler(authHandler);
+    }
+
+    void handleProviderRoute(final RoutingContext ctx, final LoraProvider provider) {
+        LOG.debug("Handling route for provider with path: [{}]", provider.pathPrefix());
+        final Span currentSpan = getCurrentSpan(ctx);
+        ctx.put(LoraConstants.APP_PROPERTY_ORIG_LORA_PROVIDER, provider.getProviderName());
+        currentSpan.setTag(TAG_LORA_PROVIDER, provider.getProviderName());
+
+        if (ctx.user() instanceof Device) {
+            final Device gatewayDevice = (Device) ctx.user();
+            final JsonObject loraMessage = ctx.getBodyAsJson();
+            currentSpan.setTag(MessageHelper.APP_PROPERTY_TENANT_ID, gatewayDevice.getTenantId());
+
+            LoraMessageType type = LoraMessageType.UNKNOWN;
+            try {
+                type = provider.extractMessageType(loraMessage);
+                final String deviceId = provider.extractDeviceId(loraMessage);
+                currentSpan.setTag(TAG_LORA_DEVICE_ID, deviceId);
+                currentSpan.setTag(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+
+                if (LoraMessageType.UPLINK.equals(type)) {
+                    final String payloadBase64 = provider.extractPayloadEncodedInBase64(loraMessage);
+                    doUpload(ctx, gatewayDevice, deviceId, payloadBase64);
+                } else {
+                    LOG.debug("Received message '{}' of type [{}] for device [{}], will discard message.", loraMessage,
+                            type, deviceId);
+                    currentSpan.log(
+                            "Received message of type '" + type + "' for device '" + deviceId + "' will be discarded.");
+                    // Throw away the message but return 202 to not cause errors on the LoRa provider side
+                    handle202(ctx);
+                }
+            } catch (final ClassCastException | LoraProviderMalformedPayloadException e) {
```

Comment:
I assume that `ClassCastException` is thrown if the payload does not contain fields of the expected type. Wouldn't it be much cleaner to also throw  `LoraProviderMalformedPayloadException` in that case?